sdk/inc/azure/core/az_mqtt5.h

```diff
@@ -0,0 +1,784 @@
+   */
+  int8_t qos;
+} az_mqtt5_pub_data;
+
+/**
+ * @brief MQTT 5 receive data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief The received properties.
+   */
+  az_mqtt5_property_bag* properties;
+
+  /**
+   * @brief The topic the message was received on.
+   */
+  az_span topic;
+
+  /**
+   * @brief The payload of the message.
+   */
+  az_span payload;
+
+  /**
+   * @brief The quality of service of the message.
+   */
+  int8_t qos;
+
+  /**
+   * @brief The ID of the message.
+   */
+  int32_t id;
+} az_mqtt5_recv_data;
+
+/**
+ * @brief MQTT 5 publish acknowledgement data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief The publish request ID.
+   */
+  int32_t id;
+} az_mqtt5_puback_data;
+
+/**
+ * @brief MQTT 5 subscribe data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief The properties of subscribe request.
+   */
+  az_mqtt5_property_bag* properties;
+  /**
+   * @brief Topic filter to subscribe to.
+   */
+  az_span topic_filter;
+
+  /**
+   * @brief Quality of service of the subscription.
+   */
+  int8_t qos;
+
+  /**
+   * @brief Id to correlate the subscription request with the response acknowledgement.
+   */
+  int32_t out_id;
+} az_mqtt5_sub_data;
+
+/**
+ * @brief MQTT 5 subscribe acknowledgement data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief The subscribe request ID.
+   */
+  int32_t id;
+} az_mqtt5_suback_data;
+
+/**
+ * @brief MQTT 5 connect data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief The properties of the connect request.
+   */
+  az_mqtt5_property_bag* properties;
+
+  /**
+   * @brief Hostname or IP address of the MQTT 5 broker.
+   */
+  az_span host;
+
+  /**
+   * @brief Port number broker is listening on.
+   */
+  int16_t port;
+
+  /**
+   * @brief The username to send to the broker.
+   */
+  az_span username;
+
+  /**
+   * @brief The password to send to the broker.
+   */
+  az_span password;
+
+  /**
+   * @brief The client ID to send to the broker.
+   */
+  az_span client_id;
+
+  /**
+   * @brief The certificate to use for authentication.
+   */
+  az_mqtt5_x509_client_certificate certificate;
+} az_mqtt5_connect_data;
+
+/**
+ * @brief MQTT 5 connect acknowledgement data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief Connection acknowledgement reason code. Indicates success or reason for failure.
+   */
+  int32_t connack_reason;
+
+  /**
+   * @brief Indicates whether a TLS authentication error occurred.
+   */
+  bool tls_authentication_error;
+} az_mqtt5_connack_data;
+
+/**
+ * @brief MQTT 5 disconnect data.
+ *
+ */
+typedef struct
+{
+  /**
+   * @brief Indicates whether a TLS authentication error occurred.
+   */
+  bool tls_authentication_error;
+
+  /**
+   * @brief Indicates whether the disconnect was requested by the client.
+   */
+  bool disconnect_requested;
+} az_mqtt5_disconnect_data;
+
+/**
+ * @brief MQTT 5 property string.
+ */
+typedef struct
+{
+  /**
+   * @brief The string value of the property.
+   */
+  az_span str;
+} az_mqtt5_property_string;
```

Comment:
It is because we want the properties to be strongly typed, it will help distinguish between `az_mqtt5_property_string` and `az_mqtt5_property_binarydata` which are both az_spans but not the same from a MQTT perspective.

Reply:
It's actually because in some MQTT implementations these need to be freed and the primitive types do not.
In some cases freeing is done by `libc/free(void*)` but in many cases it is not `somemqttstack_destroy(somemqttstack_string_type)`.